server/manager/src/main/java/org/apache/accumulo/master/Master.java

```diff
@@ -1686,4 +1741,56 @@ public boolean isActiveService() {
     return masterInitialized.get();
   }
 
+  @SuppressWarnings("deprecation")
+  void initializeBalancer() {
+
+    // Try to initialize the defined balancer as the updated TabletBalancer class first. If that
+    // fails with a ClassCastException, it means the property is specified as the deprecated
+    // balancer type, so initialize it instead. If we still end up with no balancer, then use
+    // DefaultLoadBalancer.
+    try {
+      tabletBalancer = Property.createInstanceFromPropertyName(getConfiguration(),
+          Property.TABLE_LOAD_BALANCER, TabletBalancer.class, null);
+    } catch (ClassCastException e) {
+      // ignore -- this means that the deprecated balancer type was used
+      deprecatedTabletBalancer = Property.createInstanceFromPropertyName(getConfiguration(),
+          Property.MANAGER_TABLET_BALANCER,
+          org.apache.accumulo.server.master.balancer.TabletBalancer.class, null);
+    }
```

Comment:
If the old were made to implement or extend the new SPI, then we wouldn't need to have two code paths here.

Reply:
I was curious about this and tried experimenting with it locally with the following approach.  Assuming we can make assumptions about the concrete types within Accumulo code and add needed methods to the concrete types.

```java
@Deprecated(since = "2.1.0")
public abstract class TabletBalancer implements org.apache.accumulo.core.spi.balancer.TabletBalancer {

  @Override
  public void init(BalancerEnvironment balancerEnvironment) {
    var bei = (BalancerEnvironmentImpl)balancerEnvironment;
    init(bei.getContext());
  }

  @Override
  public void getAssignments(AssignmentParameters params) {
    var api = (AssignmentParamsImpl)params; 
    Map<KeyExtent,TServerInstance> assignments = new HashMap<>();
    getAssignments(api.currentStatusOld(), api.unassignedTabletsOld(), assignments);
    assignments.forEach((ke, tsi) -> {
      params.assignmentsOut().put(new TabletIdImpl(ke), TabletServerIdImpl.fromThrift(tsi));
    });
  }

  @Override
  public long balance(BalanceParameters params) {
    var bpi = (BalanceParamsImpl) params;
    List<TabletMigration> migrationsOut = new ArrayList<>();
    balance(bpi.currentStatusOld(), bpi.currentMigrationsOld(), migrationsOut);
    migrationsOut.forEach(mo -> {
      params.migrationsOut()
          .add(new org.apache.accumulo.core.spi.balancer.data.TabletMigration(
              new TabletIdImpl(mo.tablet), TabletServerIdImpl.fromThrift(mo.oldServer),
              TabletServerIdImpl.fromThrift(mo.newServer)));
    });
  }
```

Reply:
I was just about to start working on this, but looks like you beat me to it. :) Thanks. I'll add this code change in.

Reply:
This is done in the latest commit.